bisection.py

```diff
@@ -0,0 +1,373 @@
+    --start <SHA> --end <SHA> --threshold <SCORE_THRESHOLD> \
+    --timeout <TIMEOUT_IN_MINS> --output <OUTPUT_FILE_PATH>
+"""
+
+import os
+import json
+import yaml
+import argparse
+import typing
+import re
+import subprocess
+from datetime import datetime
+from typing import Optional, List, Dict, Tuple
+
+from torchbenchmark.score.compute_score import compute_score
+from torchbenchmark.util import gitutils
+from torchbenchmark.util import torch_nightly
+
+def exist_dir_path(string):
+    if os.path.isdir(string):
+        return string
+    else:
+        raise NotADirectoryError(string)
+
+TORCH_GITREPO="https://github.com/pytorch/pytorch.git"
+TORCHBENCH_GITREPO="https://github.com/pytorch/benchmark.git"
+TORCHBENCH_SCORE_CONFIG="torchbenchmark/score/configs/v0/config-v0.yaml"
+TORCHBENCH_DEPS = {
+    "torchtext": os.path.expandvars("${HOME}/text"),
+    "torchvision": os.path.expandvars("${HOME}/vision"),
+    "torchaudio": os.path.expandvars("${HOME}/audio"),
+}
+
+class Commit:
+    sha: str
+    ctime: str
+    score: Optional[float]
+    def __init__(self, sha, ctime, score):
+        self.sha = sha
+        self.ctime = ctime
+        self.score = score
+    def __str__(self):
+        return self.sha
+
+class TorchSource:
+    srcpath: str
+    commit_date: datetime
+    commits: List[Commit]
+    # Map from commit SHA to index in commits
+    commit_dict: Dict[str, int]
+    def __init__(self, srcpath: str):
+        self.srcpath = srcpath
+        self.commits = []
+        self.commit_dict = dict()
+
+    def prep(self) -> bool:
+        repo_origin_url = gitutils.get_git_origin(self.srcpath)
+        if not repo_origin_url == TORCH_GITREPO:
+            print(f"Unmatched repo origin url: {repo_origin_url} with standard {TORCH_GITREPO}")
+            return False
+        return True
+    
+    # Get all commits between start and end, save them in commits
+    def init_commits(self, start: str, end: str) -> bool:
+        commits = gitutils.get_git_commits(self.srcpath, start, end)
+        if not commits or len(commits) < 2:
+            print(f"Failed to retrieve commits from {start} to {end} in {self.srcpath}.")
+            return False
+        for count, commit in enumerate(commits):
+            ctime = gitutils.get_git_commit_date(self.srcpath, commit)
+            self.commits.append(Commit(sha=commit, ctime=ctime, score=None))
+            self.commit_dict[commit] = count
+        # Setup commit date
+        last_commit_date = self.commits[-1].ctime.split()[0]
+        self.commit_date = datetime.strptime(last_commit_date, "%Y-%m-%d")
+        return True
+    
+    def get_mid_commit(self, left: Commit, right: Commit) -> Optional[Commit]:
+        left_index = self.commit_dict[left.sha]
+        right_index = self.commit_dict[right.sha]
+        if right_index == left_index + 1:
+            return None
+        else:
+            return self.commits[(left_index + right_index) / 2]
```

Comment:
does this work? i would think that sometimes there are an odd number divided by 2 and you get a float 1.5 index into a dict.

Reply:
Fixed. Thanks for pointing out.